datafusion-examples/examples/table_sample.rs

```diff
@@ -0,0 +1,1353 @@
+use datafusion::common::tree_node::{
+    Transformed, TransformedResult, TreeNode, TreeNodeRewriter,
+};
+
+use arrow_schema::{DataType, Field, Fields, Schema, SchemaRef, TimeUnit};
+use async_trait::async_trait;
+use datafusion::catalog::cte_worktable::CteWorkTable;
+use datafusion::common::file_options::file_type::FileType;
+use datafusion::config::ConfigOptions;
+use datafusion::datasource::file_format::format_as_file_type;
+use datafusion::datasource::{provider_as_source, DefaultTableSource, TableProvider};
+use datafusion::error::DataFusionError;
+use datafusion::execution::{
+    FunctionRegistry, SendableRecordBatchStream, SessionState, SessionStateBuilder,
+    TaskContext,
+};
+use datafusion::logical_expr::planner::{
+    ContextProvider, ExprPlanner, PlannerResult, RawBinaryExpr, TypePlanner,
+};
+use datafusion::logical_expr::sqlparser::dialect::PostgreSqlDialect;
+use datafusion::logical_expr::sqlparser::parser::Parser;
+use datafusion::logical_expr::var_provider::is_system_variables;
+use datafusion::optimizer::simplify_expressions::ExprSimplifier;
+use datafusion::optimizer::AnalyzerRule;
+use datafusion::physical_expr::EquivalenceProperties;
+use datafusion::physical_plan::execution_plan::{Boundedness, EmissionType};
+use datafusion::physical_plan::metrics::{
+    BaselineMetrics, ExecutionPlanMetricsSet, MetricsSet,
+};
+use datafusion::physical_plan::{
+    displayable, DisplayAs, DisplayFormatType, ExecutionPlan, PlanProperties,
+    RecordBatchStream,
+};
+use datafusion::physical_planner::{
+    DefaultPhysicalPlanner, ExtensionPlanner, PhysicalPlanner,
+};
+use datafusion::prelude::*;
+use datafusion::sql::planner::{ParserOptions, PlannerContext, SqlToRel};
+use datafusion::sql::sqlparser::ast::{TableSampleKind, TableSampleModifier};
+use datafusion::sql::unparser::ast::{
+    DerivedRelationBuilder, QueryBuilder, RelationBuilder, SelectBuilder,
+};
+use datafusion::sql::unparser::dialect::CustomDialectBuilder;
+use datafusion::sql::unparser::expr_to_sql;
+use datafusion::sql::unparser::extension_unparser::UserDefinedLogicalNodeUnparser;
+use datafusion::sql::unparser::extension_unparser::{
+    UnparseToStatementResult, UnparseWithinStatementResult,
+};
+use datafusion::sql::unparser::{plan_to_sql, Unparser};
+use datafusion::variable::VarType;
+use log::{debug, info};
+use std::fmt;
+use std::fmt::{Debug, Formatter};
+use std::hash::{Hash, Hasher};
+use std::pin::Pin;
+use std::str::FromStr;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+
+use arrow::array::UInt32Array;
+use arrow::compute;
+use arrow::record_batch::RecordBatch;
+
+use datafusion::execution::context::QueryPlanner;
+use datafusion::sql::sqlparser;
+use datafusion::sql::sqlparser::ast;
+use futures::stream::{Stream, StreamExt};
+use futures::TryStreamExt;
+use rand::rngs::StdRng;
+use rand::{Rng, SeedableRng};
+use rand_distr::{Distribution, Poisson};
+
+/// This example demonstrates the table sample support.
```

Comment:
I think this undersells somewhat what this example does

How about something more like this:

```suggestion
/// This example demonstrates how to extend DataFusion's SQL parser to recognize
/// other syntax.
///
/// This example shows how to extend the DataFusion SQL planner to support the
/// `TABLESAMPLE` clause in SQL queries and then use a custom user defined node
/// to implement the sampling logic in the physical plan.
```